File: 00-每日一题/20200407_3.py
import collections
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        left = 0
        right = 0
        c_dict = collections.defaultdict(int)
        max_len = 0
        s_len = len(s)
        if s_len < 2:
            return s_len

        while right < s_len:

            if c_dict[s[right]] <= 0:
                c_dict[s[right]] += 1
                right += 1

            else:
                c_dict[s[left]] -= 1
                left += 1

            if right-left > max_len:
                max_len = right-left

        return max_len


if __name__ == '__main__':

    obj = Solution()
    while True:
        s = input()
        res = obj.lengthOfLongestSubstring(s)
        print(res)

File: 00-每日一题/20200407_3.cpp
//
// Created by zanbo on 2020/4/7.
//

class Solution {
    //返回最长的无重复字符的子串长度：使用滑动窗口的方法
public:
    int lengthOfLongestSubstring(string s) {
        int myMap[256]; // 使用int数组做哈希表 不用map char一共256
        memset(myMap,-1, sizeof(myMap));
        int l=0;
        int r=0;
        int maxLen = 0;
        for(;r<s.size();r++)
        {
            if(myMap[s[r]]<l)
            {
                //之前未出现过
                // 记录目前的长度
                maxLen = max(maxLen,r-l+1);
            } else{
                // 遇到重复字符 需要移动左边界
                l = myMap[s[r]]+1;

            }
            myMap[s[r]] = r;
        }
        return maxLen;
    }
};

